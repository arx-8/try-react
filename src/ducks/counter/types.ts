export enum CounterActionType {
  ADD = "COUNTER/ADD",
  DECREMENT = "COUNTER/DECREMENT",
  INCREMENT = "COUNTER/INCREMENT",
  CHANGE_DEFAULT_AMOUNT = "COUNTER/CHANGE_DEFAULT_AMOUNT",
}
