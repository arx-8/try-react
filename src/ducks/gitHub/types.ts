import { AxiosError } from "axios"
import { User } from "types/GitHub"

export type GitHubState = Readonly<{
  users: User[]
  isLoading: boolean
  error: AxiosError | null
}>

export type FetchMembersParams = {
  companyName: string
}

export type FetchMemberResponse = Readonly<{
  users: User[]
}>

export type FetchMemberError = Readonly<{
  error: AxiosError
}>
