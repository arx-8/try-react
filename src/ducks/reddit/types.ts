import { FixMeAny } from "src/types/Utils"

export type PostsBySubredditState = {
  isFetching: boolean
  didInvalidate: boolean
  items: FixMeAny[]
}

export type SelectedSubredditState = string
