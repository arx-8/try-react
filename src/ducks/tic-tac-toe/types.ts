export enum ActionTypes {
  CLICK_SQUARE = "tic-tac-toe/CLICK_SQUARE",
  RESET = "tic-tac-toe/RESET",
}

export type InputValue = "O" | "X" | null

export type PlayerName = NonNullable<InputValue>

export type TTTState = Readonly<{
  currentPlayerName: PlayerName
  winnerPlayerName: PlayerName | null
  inputValues: InputValue[]
  isContinue: boolean
}>
