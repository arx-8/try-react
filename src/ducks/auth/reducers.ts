import { AuthToken } from "domain/models/Auth"
import produce from "immer"
import { reducerWithInitialState } from "typescript-fsa-reducers"
import { actions } from "./actions"

export type State = Readonly<{
  authToken?: AuthToken
}>

export const initialState: State = {
  authToken: undefined,
}

export const reducer = reducerWithInitialState(initialState)
  .case(actions.login, (state) => {
    return produce(state, (draft) => {
      draft.authToken = "any auth token" as AuthToken
    })
  })
  .case(actions.logout, (state) => {
    return produce(state, (draft) => {
      draft.authToken = undefined
    })
  })
  .build()
