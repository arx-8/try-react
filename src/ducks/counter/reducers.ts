import * as actions from "./actions"
import { CounterActionType } from "./types"
import { handleActions, ReduxCompatibleReducer } from "redux-actions"
import produce from "immer"

export type CounterState = Readonly<{
  count: number
  defaultAmount: number
}>

// redux-actions の型がうまく書けない。redux-actions はだいぶダメな感じ。
type Action =
  | ReturnType<
      | typeof actions.add
      | typeof actions.changeDefaultAmount
      | typeof actions.decrement
      | typeof actions.increment
    >
  | any

export const initialState: CounterState = {
  count: 0,
  defaultAmount: 1,
}

export const reducer: ReduxCompatibleReducer<
  CounterState,
  Action
> = handleActions(
  {
    [CounterActionType.ADD]: (state, action) => {
      return produce(state, (draft) => {
        draft.count = state.count + action.payload.count
      })
    },
    [CounterActionType.DECREMENT]: (state) => {
      return produce(state, (draft) => {
        draft.count = state.count - 1
      })
    },
    [CounterActionType.INCREMENT]: (state) => {
      return produce(state, (draft) => {
        draft.count = state.count + 1
      })
    },
    [CounterActionType.CHANGE_DEFAULT_AMOUNT]: (state, action) => {
      return produce(state, (draft) => {
        draft.defaultAmount = action.payload.defaultAmount
      })
    },
  },
  initialState
)
