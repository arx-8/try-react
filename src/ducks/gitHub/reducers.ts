import { Action, ActionTypes } from "./actions"
import { GitHubState } from "./types"
import { Reducer } from "redux"
import produce from "immer"

export const initialState: GitHubState = {
  users: [],
  isLoading: false,
  error: undefined,
}

export const reducer: Reducer<GitHubState, Action> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case ActionTypes.FETCH_MEMBERS_START:
      return produce(state, (draft) => {
        draft.isLoading = true
      })

    case ActionTypes.FETCH_MEMBERS_SUCCEED:
      return produce(state, (draft) => {
        const { users } = action.payload
        draft.users = users
        draft.isLoading = false
      })

    case ActionTypes.FETCH_MEMBERS_FAIL:
      return produce(state, (draft) => {
        const { error } = action.payload
        draft.error = error
        draft.isLoading = false
      })

    case ActionTypes.INITIALIZE_MEMBERS:
      return produce(state, (draft) => {
        draft.users = []
      })

    default: {
      // case の定義忘れ防止のため
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const _: never = action
      return state
    }
  }
}
