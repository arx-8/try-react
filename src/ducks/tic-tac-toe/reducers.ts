import { clickSquare } from "./actions"
import { range } from "utils/ArrayUtils"
import { reducerWithInitialState } from "typescript-fsa-reducers"
import { TTTState } from "./types"
import produce from "immer"

export const initialState: TTTState = {
  currentPlayerName: "O",
  inputValues: range(9).map(() => null),
}

export const reducer = reducerWithInitialState(initialState).case(
  clickSquare,
  (state, payload) => {
    return produce(state, (draft) => {
      draft.currentPlayerName = state.currentPlayerName === "O" ? "X" : "O"
      draft.inputValues[payload.index] = payload.value
    })
  }
)
