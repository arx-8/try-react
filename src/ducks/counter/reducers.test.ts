import { CounterActionType } from "./types"
import { counterReducer } from "."
import { initialState } from "./reducers"
import produce from "immer"

describe("CHANGE_DEFAULT_AMOUNT", () => {
  it("prev = initialState", () => {
    expect.hasAssertions()
    // ## Arrange ##
    const prevState = initialState
    // ## Act ##
    const result = counterReducer(prevState, {
      type: CounterActionType.CHANGE_DEFAULT_AMOUNT,
      payload: {
        defaultAmount: 10,
      },
    })
    // ## Assert ##
    expect(result).toMatchSnapshot()
  })

  it("prev = edited", () => {
    expect.hasAssertions()
    // ## Arrange ##
    const prevState = produce(initialState, (draft) => {
      draft.count = 10
      draft.defaultAmount = 10
    })

    // ## Act ##
    const result = counterReducer(prevState, {
      type: CounterActionType.CHANGE_DEFAULT_AMOUNT,
      payload: {
        defaultAmount: 1000,
      },
    })
    // ## Assert ##
    expect(result).toMatchSnapshot()
  })
})
