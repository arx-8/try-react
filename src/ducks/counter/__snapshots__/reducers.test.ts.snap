// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CHANGE_DEFAULT_AMOUNT prev = edited 1`] = `
Object {
  "count": 10,
  "defaultAmount": 1000,
}
`;

exports[`CHANGE_DEFAULT_AMOUNT prev = initialState 1`] = `
Object {
  "count": 0,
  "defaultAmount": 10,
}
`;
